src/Projections/Reunion/Microsoft.UI.Dispatching.DispatcherQueueSynchronizationContext.DispatcherQueueProxyHandler.cs

```diff
@@ -0,0 +1,260 @@
+            public static DispatcherQueueProxyHandler* Create(SendOrPostCallback handler, object? state)
+            {
+#if NET6_0
+                DispatcherQueueProxyHandler* @this = (DispatcherQueueProxyHandler*)NativeMemory.Alloc((nuint)sizeof(DispatcherQueueProxyHandler));
+#else
+                DispatcherQueueProxyHandler* @this = (DispatcherQueueProxyHandler*)Marshal.AllocHGlobal(sizeof(DispatcherQueueProxyHandler));
+#endif
+
+                @this->lpVtbl = Vtbl;
+                @this->callbackHandle = GCHandle.Alloc(handler);
+                @this->stateHandle = state is not null ? GCHandle.Alloc(state) : default;
+                @this->referenceCount = 1;
+
+                return @this;
+            }
+
+            /// <summary>
+            /// Devirtualized API for <c>IUnknown.Release()</c>.
+            /// </summary>
+            /// <returns>The updated reference count for the current instance.</returns>
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public uint Release()
+            {
+                uint referenceCount = Interlocked.Decrement(ref this.referenceCount);
+
+                if (referenceCount == 0)
+                {
+                    callbackHandle.Free();
+                    
+                    if (stateHandle.IsAllocated)
+                    {
+                        stateHandle.Free();
+                    }
+
+#if NET6_0
+                    NativeMemory.Free(Unsafe.AsPointer(ref this));
+#else
+                    Marshal.FreeHGlobal((IntPtr)Unsafe.AsPointer(ref this));
+#endif
+                }
+
+                return referenceCount;
+            }
+
+            /// <summary>
+            /// A private type with the implementation of the unmanaged methods for <see cref="DispatcherQueueProxyHandler"/>.
+            /// These methods will be set into the shared vtable and invoked by WinRT from the object passed to it as an interface.
+            /// </summary>
+            private static class Impl
+            {
+                /// <summary>
+                /// The HRESULT for a successful operation.
+                /// </summary>
+                private const int S_OK = 0;
+
+                /// <summary>
+                /// The HRESULT for an invalid cast from <c>IUnknown.QueryInterface</c>.
+                /// </summary>
+                private const int E_NOINTERFACE = unchecked((int)0x80004002);
+
+                /// <summary>
+                /// The GUID for the <c>IUnknown</c> COM interface.
+                /// </summary>
+                private static ref readonly Guid GuidOfIUnknown
+                {
+                    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+                    get
+                    {
+                        ReadOnlySpan<byte> data = new byte[] { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xC0, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x46 };
```

Comment:
What does guarantee that the memory is properly aligned?

Multiple people just spent a week tracking a bug in dotnet/runtime where the C/C++ compiler was assuming proper memory alignment, and producing wrong code due to not it being the case: https://github.com/dotnet/runtime/issues/61671

Reply:
This project is using standard readonly Guids constants in many other places so using this hack in a few places does not really move the needle. And the key reason why terrafx is using this hack (single type with thousands of Guid constants, most of them never used) does not apply here.

Reply:
This trick originally came up while investigating some performance issues in Paint.NET, as @rickbrew was noticing a big startup slowdown due to a static constructor initializing a few thousand GUIDs. @tannergooding thought about using this to fix that, and it's now part of `TerraFX.Interop.Windows` as well (which is then used by projects such as Paint.NET, ComputeSharp, etc.). IIRC we had a very similar conversation before (other folks raised the point about unaligned reads potentially being a problem, especially on ARM64), and going back to older messages on Discord:

> "Btw does windows support arm CPUs without unaligned reads"

> Tanner: "Don't believe so. Arm64 (like x64) always supports unaligned reads (well outside a special switch which I don't think most apps will work with) and windows doesn't support standalone arm32, just the arm32 subsystem of arm64 which is like v8, rather than v7)"

I think I recall the conclusion being that doing this in this case (and x86/64 doesn't care anyway), and also in a scenario where you only need to run on Windows (such as CsWinRT) and therefore having the same hardware requirements that Windows has), this is fine. That said if that's not the case I'm happy to revert this, and I guess then we would also have a problem in TerraFX then though. Either way it seems we'll (or at least, I'll) learn something new today 😄

Reply:
The problem in dotnet/runtime#61671 was that the ABI cares, nothing to do with whether the CPU cares.

Reply:
> What does guarantee that the memory is properly aligned?

@jkotas, if the compiler (C#, JIT, Crossgen, etc) isn't ensuring that RVA statics are at least 4-byte aligned already then we likely already have issues with `RuntimeHelpers.InitializeArray` and will run into issues with recently approved `RuntimeHelpers.CreateSpan`.

Likewise, if the underlying ABI expects `Guid` to be more than 4-byte aligned then .NET already has issues because that's the natural alignment we have for `Guid`, not `8` and not `16`.

That isn't necessarily a good reason to introduce risk here, particularly if there are only a handful of `guids` that are trivial to initialize; but it should mean that this pattern is "safe" for up to `int`

Reply:
> If that's not the case

Looks like the C# compiler tightly packs the data, so if you have:
```
public ReadOnlySpan<byte> M() {
    return new byte[] { 1, 2, 3 };
}

public uint[] N() {
    return new uint[] { 1, 2, 3 };
}
```

Then the backing RVA for `new uint[]` (the one currently passed to `InitializeArray` and that will be passed to `CreateSpan` in the future) may be unaligned.

Reply:
> Then the backing RVA for new uint[] (the one currently passed to InitializeArray and that will be passed to CreateSpan in the future) may be unaligned.

Good catch. It needs to be taken care of as part the part of the Roslyn enablement of `RuntimeHelpers.CreateSpan`. The block has to have same alignment as the element. But I do not think it makes sense to make Roslyn to apply 4 or 8 byte alignment to ReadOnlySpan<byte> constants.

FWIW, InitializeArray does not require the block to be aligned. It is fine since it is one-time operation. The (currently unused) big endian path for InitializeArray is using unaligned reads for this reason: https://github.com/dotnet/runtime/blob/3efb476e2e22f7dd5ac96968203ef931704d34b8/src/coreclr/classlibnative/bcltype/arraynative.cpp#L1134-L1142

> As I mentioned I have no issue reverting this, to me it was more a matter of: is this guaranteed to be safe or not?

I would qualify it as fragile code. The spec does not guarantee that it will work. It works today, but it may stop working as things evolve.

Reply:
I'll go revert the change then, we can always change this in another PR later down the line anyway. This whole conversation was rally interesting though, so definitely worth the time, especially if we apparently also found a potential issue with Roslyn and the way we're currently relying on this while investigating this. Thanks everyone for chiming in! 😄